xapian-letor/api/featurelist_internal.cc

```diff
@@ -0,0 +1,177 @@
+	Xapian::doccount df = featurelist_db.get_termfreq(*qt);
+	idf[*qt] = (df == 0) ? 0 : log10((double)totaldocs / (double)(1 + df));
+    }
+    std::swap(inverse_doc_freq, idf);
+}
+
+void
+FeatureList::Internal::compute_doc_length()
+{
+    std::map<std::string, long int> len;
+
+    long int temp_count = 0;
+    Xapian::TermIterator dt = featurelist_doc.termlist_begin();
+    // reach the iterator to the start of the title terms i.e. prefix "S"
+    dt.skip_to("S");
+    for ( ; dt != featurelist_doc.termlist_end(); ++dt) {
+	if ((*dt)[0] != 'S') {
+	    // We've reached the end of the S-prefixed terms.
+	    break;
+	}
+	temp_count += dt.get_wdf();
+    }
+    len["title"] = temp_count;
+    len["whole"] = featurelist_db.get_doclength(featurelist_doc.get_docid());
+    len["body"] = len["whole"] - len["title"];
+    std::swap(doc_length, len);
+}
+
+void
+FeatureList::Internal::compute_collection_length()
+{
+    std::map<std::string, long int> len;
+
+    if (!featurelist_db.get_metadata("collection_len_title").empty() &&
+	    !featurelist_db.get_metadata("collection_len_body").empty() &&
+	    !featurelist_db.get_metadata("collection_len_whole").empty()) {
+	len["title"] = atol(featurelist_db.get_metadata("collection_len_title").c_str());
+	len["body"] = atol(featurelist_db.get_metadata("collection_len_body").c_str());
+	len["whole"] = atol(featurelist_db.get_metadata("collection_len_whole").c_str());
+    } else {
+	long int temp_count = 0;
+	Xapian::TermIterator dt = featurelist_db.allterms_begin("S");
+	for ( ; dt != featurelist_db.allterms_end("S"); ++dt) {
+	    //  because we don't want the unique terms so we want their
+	    // original frequencies and i.e. the total size of the title collection.
+	    temp_count += featurelist_db.get_collection_freq(*dt);
+	}
+	len["title"] = temp_count;
+	len["whole"] = featurelist_db.get_avlength() * featurelist_db.get_doccount();
+	len["body"] = len["whole"] - len["title"];
+    }
+    std::swap(collection_length, len);
+}
+
+void
+FeatureList::Internal::compute_collection_termfreq()
+{
+    std::map<std::string, long int> tf;
+
+    for (Xapian::TermIterator qt = featurelist_query.get_unique_terms_begin();
+	 qt != featurelist_query.get_terms_end(); ++qt) {
+	tf[*qt] = featurelist_db.get_collection_freq(*qt);
+    }
+    std::swap(collection_termfreq, tf);
+}
+
+void
+FeatureList::Internal::populate_feature(Feature *feature_)
+{
+    stat_flags stats_needed = stat_flags(feature_->get_stats());
+    if (stats_needed & TERM_FREQUENCY)
+	feature_->set_termfreq(termfreq);
+    if (stats_needed & INVERSE_DOCUMENT_FREQUENCY)
+	feature_->set_inverse_doc_freq(inverse_doc_freq);
+    if (stats_needed & DOCUMENT_LENGTH)
+	feature_->set_doc_length(doc_length);
+    if (stats_needed & COLLECTION_LENGTH)
+	feature_->set_collection_length(collection_length);
+    if (stats_needed & COLLECTION_TERM_FREQ)
+	feature_->set_collection_termfreq(collection_termfreq);
```

Comment:
There's clearly some part of this I'm not explaining to you well enough, but I'm not sure what part that is.

The complexity guarantees of of `std::map::clear()` and `std::map::~map()` are both linear:

http://en.cppreference.com/w/cpp/container/map/clear :
> Complexity
>
> Linear in the size of the container.

http://en.cppreference.com/w/cpp/container/map/%7Emap :
> Complexity
>
> Linear in the size of the container.

But if the container has already been cleared, then N=0 for the destructor (similarly `clear()` on an already empty container is with N=0).

So calling `clear()` does **NOT** incur extra work, it just does the work which has to be done at some point.  If we don't explicitly do it, it'll have to happen when the destructor gets called.  The only way to avoid doing that work is to avoid calling the destructor, which would require leaking the map and its contents, or causing the process to exit uncleanly (e.g. via `std::_Exit()` or an uncaught exception).

> since all the stats are required by the features, calculating them at the beginning or when the stat is computed for the first time wouldn't affect the time complexity at all

If this is true, then the optimisation won't help, but not for the reason you keep insisting.

But doesn't this depend on which features are being worked with?